lib/thredded/database_seeder.rb

```diff
@@ -0,0 +1,269 @@
+      ]
+      log "Creating #{additional_messageboards.length} additional messageboards..."
+      additional_messageboards.each do |(name, description, group_id)|
+        messageboard = Messageboard.create!(name: name, description: description, messageboard_group_id: group_id)
+        FactoryGirl.create_list(:topic, 1 + rand(3), messageboard: messageboard, with_posts: 1)
+      end
+    end
+
+    def topics(count: 1)
+      @topics ||= Topics.new(self).find_or_create(count: count)
+    end
+
+    def private_topics(count: 1)
+      @private_topics ||= PrivateTopics.new(self).find_or_create(count: count)
+    end
+
+    def posts(count: (1..1))
+      @posts ||= Posts.new(self).find_or_create(count: count)
+    end
+
+    def private_posts(count: (1..1))
+      @private_posts ||= PrivatePosts.new(self).find_or_create(count: count)
+    end
+
+    def follow_some_topics(count: (5..10), count_users: (1..5))
+      log 'Following some topics...'
+      posts.each do |post|
+        Thredded::UserTopicFollow.create_unless_exists(post.user_id, post.postable_id, :posted) if post.user_id
+      end
+      follow_some_topics_by_user(first_user, count: count)
+      users.sample(count_users.min + rand(count_users.max - count_users.min + 2)).each do |user|
+        follow_some_topics_by_user(user, count: count)
+      end
+    end
+
+    def follow_some_topics_by_user(user, count: (1..10))
+      topics.sample(count.min + rand(count.max - count.min + 2)).each do |topic|
+        Thredded::UserTopicFollow.create_unless_exists(user.id, topic.id)
+      end
+    end
+
+    def read_some_topics(count: (5..10), count_users: (1..5))
+      log 'Reading some topics...'
+      topics.each do |topic|
+        read_topic(topic, topic.last_user_id) if topic.last_user_id
+      end
+      read_some_topics_by_user(first_user, count: count)
+      @users.sample(count_users.min + rand(count_users.max - count_users.min + 2)).each do |user|
+        read_some_topics_by_user(user, count: count)
+      end
+    end
+
+    def read_some_topics_by_user(user, count: (1..10))
+      topics.sample(count.min + rand(count.max - count.min + 2)).each do |topic|
+        read_topic(topic, user.id)
+      end
+    end
+
+    def read_topic(topic, user_id)
+      Thredded::UserTopicReadState.find_or_initialize_by(user_id: user_id, postable_id: topic.id)
+        .update!(read_at: topic.updated_at, page: 1)
+    end
+
+    class BaseSeedData
+      attr_reader :seeder
+
+      def initialize(seed_database)
+        @seeder = seed_database
+      end
+
+      delegate :log, to: :seeder
+
+      def model_class
+        self.class::MODEL_CLASS
+      end
+
+      delegate :exists?, to: :model_class
+
+      def create(*_args)
+        fail 'override in your class!'
+      end
+
+      def find
+        model_class.all.to_a
+      end
+
+      def find_or_create(*args)
+        return @stored if @stored
+        @stored = if exists?
```

Comment:
great, but don't need to use `presence` at all if find returns `nil`. Not 100% sure if I need to do the to_a to ensure array is cached rather than re-returned, but it works.